system/metrics/sinks/prometheus/prometheusReporterService.cpp

```diff
@@ -0,0 +1,354 @@
+            e->Release();
+        }
+        catch (...)
+        {
+            IERRLOG("PrometheusMetricsService: HandleReq: Unknown exception occurred");
+        }
+    }
+
+    LOG(MCuserInfo, "PrometheusMetricsService: stopping - port: '%i'\n", m_port);
+
+    return 0;
+}
+
+bool PrometheusReporterService::onException(CHttpResponse * response, const char * message)
+{
+    IERRLOG("PrometheusReporterService encountered exception: %s", message);
+
+    HtmlPage page(HTTP_PAGE_TITLE);
+    page.appendContent(new CHtmlHeader(H1, "Internal Error"));
+
+    if (message && *message)
+        page.appendContent(new CHtmlText(message));
+    else
+        page.appendContent(new CHtmlText("Unknown Error"));
+
+    StringBuffer content;
+    page.getHtml(content);
+
+    response->setContent(content.length(), content.str());
+    response->setStatus(HTTP_STATUS_INTERNAL_SERVER_ERROR);
+
+    response->send();
+
+    return true;
+}
+
+bool PrometheusReporterService::onException(CHttpResponse * response, IException * e)
+{
+    StringBuffer errmessage;
+    if (e)
+        e->errorMessage(errmessage);
+    else
+        errmessage.set("Unknown Exception Error");
+
+    return onException(response, errmessage.str());
+}
+
+bool PrometheusReporterService::onGetNotFound(CHttpResponse * response, const char * path)
+{
+    return onMethodNotFound(response, "GET", path);
+}
+
+bool PrometheusReporterService::onPostNotFound(CHttpResponse * response, const char * path)
+{
+    return onMethodNotFound(response, "POST", path);
+}
+
+bool PrometheusReporterService::onMethodNotFound(CHttpResponse * response, const char * httpMethod, const char * path)
+{
+    HtmlPage page(HTTP_PAGE_TITLE);
+    VStringBuffer message("Target Service Not Found: '%s' '%s'", httpMethod, path);
+    page.appendContent(new CHtmlHeader(H1, message.str()));
+
+    StringBuffer content;
+    page.getHtml(content);
+
+    response->setContent(content.length(), content.str());
+    response->setStatus(HTTP_STATUS_NOT_FOUND);
+
+    response->send();
+    return true;
+}
+
+bool PrometheusReporterService::onMetricsUnavailable(CHttpRequest * request, CHttpResponse * response)
+{
+    HtmlPage page(HTTP_PAGE_TITLE);
+    VStringBuffer message("Metrics are not available: '%s'", request->queryPath());
+    page.appendContent(new CHtmlHeader(H1, message.str()));
+
+    StringBuffer content;
+    page.getHtml(content);
+
+    response->setContent(content.length(), content.str());
+    response->setStatus(HTTP_STATUS_OK);
+
+    response->send();
+    return true;
+}
+
+const char * mapMockMetricTypeToPrometheusStr(MetricType type)
+{
+    switch (type)
+    {
+        case hpccMetrics::METRICS_COUNTER:
+            return "counter";
+        case hpccMetrics::METRICS_GAUGE:
+            return "gauge";
+        default:
+            WARNLOG("Encountered unknown metric - cannot map to Prometheus metric!");
```

Comment:
This could potentially generate a lot of logging.

Reply:
I'm conflicted about this one...
@kenrowland is there a way for the sink to query the framework for metrics to be reported at init time?
I know metrics can be added dynamically, but this would allow the sink to report potential problems once at init time, with out needed to keep track of metrics, and performing look-ups at runtime.